### ISSUE TITLE
Missing log information after the pre-decoding phase

### ISSUE THREAD

**Issue by Nicogp** [wazuh/wazuh#25536]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.8.2 | wazuh-analysisd | Manager | Packages/Sources | OS version |

Hi Team,

A community user reported that after upgrading from 4.4.4 to 4.8.2 some of his custom decoders stopped working.

<details><summary>decoders</summary>

```xml
<decoder name="x_Server_Decode">
    <program_name>local_decoder_Wazuh</program_name>
    </decoder>

<decoder name="x_Server_INFO">
<prematch>^INFO  [</prematch>
</decoder>

<decoder name="x_Server_WARN">
<prematch>^WARN  [</prematch>
</decoder>

<decoder name="x_Server_ERROR">
<prematch>^ERROR [</prematch>
</decoder>

<decoder name="INfo_IdentityID">
  <parent>x_Server_INFO</parent>
  <regex>^(\.+)  (\.+)</regex>
  <order>state,description</order>
</decoder>

<decoder name="Warn_IdentityID">
  <parent>x_Server_WARN</parent>
  <regex>^(\.+)  (\.+)</regex>
  <order>state,description</order>
</decoder>

<decoder name="Error_IdentityID">
  <parent>x_Server_ERROR</parent>
  <regex>^(\.+) (\.+)</regex>
  <order>state,description</order>
</decoder>
```

</details>

<details><summary>wazuh-logtest output (4.4.4)</summary>

```
# /var/ossec/bin/wazuh-logtest 
Starting wazuh-logtest v4.4.4
Type one log per line

2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.

**Phase 1: Completed pre-decoding.
        full event: '2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
        timestamp: '2024-07-16 11:07:19,98'

**Phase 2: Completed decoding.
        name: 'x_Server_INFO'
        description: '[com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
        state: 'INFO'
```

</details>

I have done some tests and I have noticed that in 4.8.2 after the predecoding phase a little bit of log information has been lost:

<details><summary>decoders</summary>

```xml
<decoder name="x_Server_INFO">
<prematch>\.+</prematch>
</decoder>

<decoder name="INfo_IdentityID">
  <parent>x_Server_INFO</parent>
  <regex>^(\.+)</regex>
  <order>rest</order>
</decoder>
```

</details>

wazuh-logtest output:

```
# /var/ossec/bin/wazuh-logtest 
Starting wazuh-logtest v4.8.2
Type one log per line

2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.

**Phase 1: Completed pre-decoding.
        full event: '2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
        timestamp: '2024-07-16 11:07:19,980'

**Phase 2: Completed decoding.
        name: 'x_Server_INFO'
        rest: ' [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
```

As we can see in the previous output the part `INFO ` of the log has been lost.

If we repeat the same test in 4.4.4, we can see that this does not occur:

```
# /var/ossec/bin/wazuh-logtest 
Starting wazuh-logtest v4.4.4
Type one log per line

2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.

**Phase 1: Completed pre-decoding.
        full event: '2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
        timestamp: '2024-07-16 11:07:19,98'

**Phase 2: Completed decoding.
        name: 'x_Server_INFO'
        rest: 'INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
```

**Comment by Dwordcito** [comment]
@wazuh/devel-cppserver-div3 please RCA and check if this issue is a regression in the analysisd functionality.

**Comment by juliancnn** [comment]
## Update

I confirm that this bug was not present in tag `v4.5.4` but it is present in the following tag `v4.6.0`.
I have also reviewed the code and tested to find where it was entered. PR https://github.com/wazuh/wazuh/pull/15826 introduces the bug as part of a fix of a similar log, testing in 4.6 and backtracking that change you can see that it works correctly. While the tests were added for the new log, there is no test for the issue log as these tests were introduced much later than the code, making it impossible to detect the corner case.



### Evidence

#### :green_circle: 4.5.4
```bash
╰─# /var/ossec/bin/wazuh-logtest
Starting wazuh-logtest v4.5.4
Type one log per line

2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.

**Phase 1: Completed pre-decoding.
        full event: '2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
        timestamp: '2024-07-16 11:07:19,98'

**Phase 2: Completed decoding.
        name: 'x_Server_INFO'
        rest: 'INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
```

#### :red_circle: 4.6.0
```bash
╰─# /var/ossec/bin/wazuh-logtest
Starting wazuh-logtest v4.6.0
Type one log per line

2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.

**Phase 1: Completed pre-decoding.
        full event: '2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'
        timestamp: '2024-07-16 11:07:19,980'

**Phase 2: Completed decoding.
        name: 'x_Server_INFO'
        rest: ' [com.identity.fsm.SessionTimeout] Application '/render' session timeout is 480 minutes.'

```

### Code seccion of interest

https://github.com/wazuh/wazuh/blob/fedb84f73b3a784e9b0c0d9e68e2692883707153/src/analysisd/cleanevent.c#L112-L134

**Comment by juliancnn** [comment]
Sure, here is the translated and improved version of your update comment for the GitHub issue:

## Update

The issue appears to be more complex than initially thought, particularly in terms of its scope.

The fix at line 112 progresses to after the timestamp, correcting the log's position and ensuring the timestamp is accurate, meaning the last digit isn't lost.

https://github.com/wazuh/wazuh/blob/fedb84f73b3a784e9b0c0d9e68e2692883707153/src/analysisd/cleanevent.c#L112

According to the syslog standard, there should be a space between the timestamp and the hostname, and another space between the hostname and the syslog payload, which may or may not include a program_name (tag) and PID. This can be verified with the [RFC 3164](https://www.ietf.org/rfc/rfc3164.txt) in section 4.1.2

```plaintext
...
A single space character MUST follow the TIMESTAMP field.
...
```
In practice, we know that there can be one or two spaces between the hostname and the timestamp, so it's challenging to distinguish between `<timestamp> <hostname> <syslog message>` and `<timestamp> <syslog message>`. The timestamp fix aligns with the standard, and `INFO` becomes the hostname.

This can be validated through the following rules, which clearly match the hostname field with `INFO` or another value:

```xml
<!-- 
2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.
2024-07-16 11:07:19,980  OtherHostaname  [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.
-->
  <rule id="100002" level="5">
    <decoded_as>x_Server_INFO</decoded_as>
    <hostname>INFO</hostname>
    <description>decoded hostname: $(hostname)</description>
  </rule>

  <rule id="100003" level="5">
    <decoded_as>x_Server_INFO</decoded_as>
    <hostname negate="yes">INFO</hostname>
    <description>decoded hostname: $(hostname)</description>
  </rule>

```

#### Logtest output

```console
╰─# /var/ossec/bin/wazuh-logtest                                                                                                                                             130 ↵
Starting wazuh-logtest v4.9.0
Type one log per line

2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.

**Phase 1: Completed pre-decoding.
        full event: '2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.'
        timestamp: '2024-07-16 11:07:19,980'

**Phase 2: Completed decoding.
        name: 'x_Server_INFO'
        rest: ' [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.'

**Phase 3: Completed filtering (rules).
        id: '100002'
        level: '5'
        description: 'decoded hostname: INFO'
        groups: '['local', 'syslog', 'sshd']'
        firedtimes: '1'
        mail: 'False'
**Alert to be generated.

2024-07-16 11:07:19,980  otherHostname  [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.

**Phase 1: Completed pre-decoding.
        full event: '2024-07-16 11:07:19,980  otherHostname  [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.'
        timestamp: '2024-07-16 11:07:19,980'

**Phase 2: Completed decoding.
        name: 'x_Server_INFO'
        rest: ' [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.'

**Phase 3: Completed filtering (rules).
        id: '100003'
        level: '5'
        description: 'decoded hostname: otherHostname'
        groups: '['local', 'syslog', 'sshd']'
        firedtimes: '1'
        mail: 'False'
**Alert to be generated.
```

While the hostname exists, it is stored and can be used in rules; however, it is not displayed in wazuh-logtest nor stored in the alert JSON. This is due to the following code section:

https://github.com/wazuh/wazuh/blob/d3e254e3a9219bbeb6a844259144257883d8f02e/src/analysisd/format/json_extended.c#L432-L445

Where the hostname is sought in the syslog message through a regex that only supports one standard.

This was introduced as part of a fix in wazuh v3.6: PR [invalid location and hostname in alerts #1213](https://github.com/wazuh/wazuh/pull/1213)

## Conclusion

- The fix is correct, and the behavior is expected. The hostname is stored and can be used in rules but is not displayed.
- The behavior of wazuh-logtest is correct and can be validated through rules.
- While `INFO` is not a valid hostname, it is a valid value for the syslog hostname field, so it's impossible to differentiate
  between `<timestamp> <hostname> <syslog message>` and `<timestamp> <syslog message>`.
- It is not advisable to change the fix again, as this could break rules based on the current behavior.

### Workaround

Users experiencing this issue can follow an alternative solution. It is possible to modify the log at the agent level before sending it to the manager to conform to the syslog standard. This can be done using the `out_format` option ([documentation here](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/localfile.html?highlight=out_format#out-format)). This option could be used when collecting the log, as shown in the example below:

```xml
<localfile>
    <location>/app/application.log</location>
    <log_format>syslog</log_format>
    <out_format>$(timestamp) $(hostname) custom_program_name: $(log)</out_format>
</localfile>
```
This will ensure that the collected log is prefixed with a syslog header, adjusting it to the standard, for example:
```plaintext
Sep 04 14:53:08 hostNameAgent test_program_name: 2024-07-16 11:07:19,980  INFO  [com.identity.fsm.SessionTimeout] Application /render session timeout is 480 minutes.
```

This allows for the creation of a decoder like the following:

```xml
 <decoder name="custom_program_name">
     <program_name>^custom_program_name</program_name>
  </decoder>
```

Then you can have rules like the following:
```xml
<group name="application-logs">
  <rule id="100002" level="5">
    <decoded_as>custom_program_name</decoded_as>
    <description>Application log message</description>
  </rule>
</group>
```

This should be sufficient to have rules and decoders capable of processing logs that do not comply with the syslog standard.

**Comment by Dwordcito** [comment]
LGTM, workaround proposed.
